### ISSUE TITLE
T-SQL does not allow COLLATE in JOIN ON clauses

### ISSUE THREAD

**Issue by borchero** [sqlfluff/sqlfluff#3671]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Joins with `COLLATE` expressions cannot be parsed with T-SQL dialect.

### Expected Behaviour

The query should be parseable without any issue (see [documentation](https://docs.microsoft.com/en-us/sql/t-sql/statements/collations?view=sql-server-ver16)).

### Observed Behaviour

Output of `sqlfluff lint <file>`:

```
== [collate.sql] FAIL                                                                                                                                 
L:   4 | P:  32 |  PRS | Line 4, Position 32: Found unparsable section: 'COLLATE                                                                      
                       | Latin1_GENERAL_CS_AS;\n'
WARNING: Parsing errors found and dialect is set to 'tsql'. Have you configured your dialect correctly?
All Finished 📜 🎉!
```

### How to reproduce

```sql
SELECT table1.col
FROM table1
INNER JOIN table2
    ON table1.col = table2.col COLLATE Latin1_GENERAL_CS_AS;
```

### Dialect

T-SQL

### Version

- `sqlfluff, version 1.2.1` (installed via homebrew)

### Configuration

```
[sqlfluff]
dialect = tsql
```

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
